saffron/src/storage_proof.rs

```diff
@@ -0,0 +1,324 @@
+        fq_sponge.squeeze(2)
+    };
+
+    let fq_sponge_before_evaluations = fq_sponge.clone();
+    let mut fr_sponge = CurveFrSponge::new(Curve::sponge_params());
+    fr_sponge.absorb(&fq_sponge.digest());
+
+    // TODO: check and see if we need to also absorb the absorb the poly cm
+    // see https://github.com/o1-labs/proof-systems/blob/feature/test-data-storage-commitments/data-storage/src/main.rs#L265-L269
+    fr_sponge.absorb(&proof.combined_data_eval);
+
+    srs.verify(
+        group_map,
+        &mut [BatchEvaluationProof {
+            sponge: fq_sponge_before_evaluations,
+            evaluation_points: vec![evaluation_point],
+            polyscale: ScalarField::one(),
+            evalscale: ScalarField::one(),
+            evaluations: vec![Evaluation {
+                commitment: PolyComm {
+                    chunks: vec![combined_data_commitment],
+                },
+                evaluations: vec![vec![proof.combined_data_eval]],
+            }],
+            opening: &proof.opening_proof,
+            combined_inner_product: proof.combined_data_eval,
+        }],
+        rng,
+    )
+}
+
+#[instrument(skip_all, level = "debug")]
+pub fn verify(
+    srs: &SRS<Curve>,
+    group_map: &<Curve as CommitmentCurve>::Map,
+    commitments: &[Curve],
+    challenge: ScalarField, // this could be merkle tree root
+    proof: &StorageProof,
+    rng: &mut OsRng,
+) -> bool {
+    let powers = commitments
+        .iter()
+        .scan(ScalarField::one(), |acc, _| {
+            let res = *acc;
+            *acc *= challenge;
+            Some(res.into_bigint())
+        })
+        .collect::<Vec<_>>();
+
+    // randomised data commitment is ∏ C_i^{chal^i} for all chunks
+    let combined_data_commitment =
+        ProjectiveCurve::msm_bigint(commitments, powers.as_slice()).into_affine();
+
+    verify_wrt_combined_data_commitment(srs, group_map, combined_data_commitment, proof, rng)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        commitment::{combine_commitments, commit_to_field_elems},
+        env,
+        utils::{encode_for_domain, test_utils::UserData},
+    };
+
+    use ark_ff::UniformRand;
+    use ark_poly::{EvaluationDomain, Radix2EvaluationDomain};
+    use kimchi::groupmap::GroupMap;
+    use mina_curves::pasta::{Fp, Vesta};
+    use once_cell::sync::Lazy;
+    use poly_commitment::{commitment::CommitmentCurve, ipa::SRS, SRS as _};
+    use proptest::prelude::*;
+
+    static SRS: Lazy<SRS<Vesta>> = Lazy::new(|| {
+        if let Ok(srs) = std::env::var("SRS_FILEPATH") {
+            env::get_srs_from_cache(srs)
+        } else {
+            SRS::create(1 << 16)
+        }
+    });
+
+    static DOMAIN: Lazy<Radix2EvaluationDomain<Fp>> =
+        Lazy::new(|| Radix2EvaluationDomain::new(SRS.size()).unwrap());
+
+    static GROUP_MAP: Lazy<<Vesta as CommitmentCurve>::Map> =
+        Lazy::new(<Vesta as CommitmentCurve>::Map::setup);
+
+    proptest! {
```

Comment:
I would be in favor of not using `proptest` anymore, as it does not provide any reproducible environment. It created some issues in the past.